File: trunk/product/src/control/DC0069.php
<?php
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0069 Controller: Contacts
 */

require_once(PATH_CTRL_FRMWK."/Controller.php");

class DC0069 extends Controller {


private function preQuery(&$params, &$where) {
    if (!empty($_REQUEST["QRY_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.ID like :ID";
      $params["ID"] = $_REQUEST["QRY_ID"];
    }
    if (!empty($_REQUEST["QRY_BPARTNER_ID"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.BPARTNER_ID like :BPARTNER_ID";
      $params["BPARTNER_ID"] = $_REQUEST["QRY_BPARTNER_ID"];
    }
    if (!empty($_REQUEST["QRY_FIRSTNAME"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.FIRSTNAME like :FIRSTNAME";
      $params["FIRSTNAME"] = $_REQUEST["QRY_FIRSTNAME"];
    }
    if (!empty($_REQUEST["QRY_LASTNAME"])) {
      $where .= (!empty($where))?" and ":"";
      $where .= "c.LASTNAME like :LASTNAME";
      $params["LASTNAME"] = $_REQUEST["QRY_LASTNAME"];
    }
}

public function doQuery() {
  try {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                c.ID
                ,c.BPARTNER_ID
                ,c.NAME
                ,c.FIRSTNAME
                ,c.LASTNAME
                ,c.PHONE
                ,c.EMAIL
                ,c.FAX
                ,c.NOTES
                ,c.MOBILE
                ,c.CREATEDON
                ,c.CREATEDBY
                ,c.MODIFIEDON
                ,c.MODIFIEDBY
                ,(select bp.name from bpartner bp where bp.id = c.bpartner_id) BPARTNER_NAME
            from BP_CONTACT c $where $orderByClause ";
    $rs = $this->db->SelectLimit($sql, $limit, $start, $params);
    $rsCount = $this->db->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $columns = array(
      "ID"
      ,"BPARTNER_ID"
      ,"NAME"
      ,"FIRSTNAME"
      ,"LASTNAME"
      ,"PHONE"
      ,"EMAIL"
      ,"FAX"
      ,"NOTES"
      ,"MOBILE"
      ,"CREATEDON"
      ,"CREATEDBY"
      ,"MODIFIEDON"
      ,"MODIFIEDBY"
      ,"BPARTNER_NAME"
      );
    $dataOut = $this->serializeCursor($rs,$columns, $this->query_data_format);
    if ($this->query_data_format == "xml" ) {header("Content-type: application/xml");}
    print $this->formatQueryResponseData($dataOut, $rsCount->fields["TOTALCOUNT"]);
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doQuery  */


public function doExport() {
  try {
    $start = nvl($this->getRequestParam("start"), 0);
    $limit = nvl($this->getRequestParam("limit"),20);
    $groupBy = (!empty($_REQUEST["groupBy"]))?$_REQUEST["groupBy"]:"";
    $orderBy = (!empty($_REQUEST["sort"]))?$_REQUEST["sort"]:"";
    $orderSense = (!empty($_REQUEST["dir"]))?$_REQUEST["dir"]:"";
    $orderByClause = (!empty($orderBy))? "order by $orderBy $orderSense " : "" ;
    $where = "";
    $params = array();
    $this->preQuery($params, $where);
    if (!empty($where)) {
      $where = " where ".$where;
    }
    $sql = "select 
                c.ID
                ,c.BPARTNER_ID
                ,c.NAME
                ,(select bp.name from bpartner bp where bp.id = c.bpartner_id) BPARTNER_NAME
                ,c.FIRSTNAME
                ,c.LASTNAME
                ,c.PHONE
                ,c.EMAIL
                ,c.FAX
                ,c.NOTES
                ,c.MOBILE
                ,c.CREATEDON
                ,c.CREATEDBY
                ,c.MODIFIEDON
                ,c.MODIFIEDBY
            from BP_CONTACT c $where $orderByClause ";
    $rs = $this->db->Execute($sql, $params);
    $rsCount = $this->db->Execute("select count(*) TOTALCOUNT from (".$sql.") t", $params);
    $rsCount->MoveFirst();
    $columns = array(
     "ID"
     ,"BPARTNER_ID"
     ,"NAME"
     ,"BPARTNER_NAME"
     ,"FIRSTNAME"
     ,"LASTNAME"
     ,"PHONE"
     ,"EMAIL"
     ,"FAX"
     ,"NOTES"
     ,"MOBILE"
     ,"CREATEDON"
     ,"CREATEDBY"
     ,"MODIFIEDON"
     ,"MODIFIEDBY"
      );
    if (!empty($_REQUEST["_p_disp_cols"])) {
      $columns = explode("|",$_REQUEST["_p_disp_cols"]);
    }
    $dataOut = $this->serializeCursor($rs,$columns,"xml");
    $dataOut = "<records>".$dataOut."</records>";
    $dataOut = "<queryParams>".$this->serializeArray($params,"xml")."</queryParams>".$dataOut;
    $dataOut = "<columnDef>".$this->columnDefForExport($columns,$this->fieldDef,true).$this->columnDefForExport(array_diff(array_keys($params), $columns),$this->fieldDef,false)."</columnDef>".$dataOut;
    $dataOut = "<staticText>".$this->exportLocalizedStaticText()."</staticText>".$dataOut;
    $dataOut = "<groupBy>".$groupBy."</groupBy>".$dataOut;
    $dataOut = "<reportData  title=\"".$this->getDcTitle()."\" by=\"".$_SESSION["user"]["userName"]."\" on=\"".date(DATE_FORMAT)."\">".$dataOut."</reportData>";
    $this->beginExport();
    print $dataOut;
    $this->endExport();
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doExport  */


public function fetchRecord() {
  try {
    $RECORD = array();
    $RECORD["ID"] = $this->getRequestParam("ID");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0069.fetchRecord().");}
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function fetchRecord */


public function doInsert() {
  $this->logger->debug("start: ".$this->dcName.".doInsert");
  try {
    $RECORD = array();
    $RECORD["BPARTNER_ID"] = $this->getRequestParam("BPARTNER_ID");
    $RECORD["BPARTNER_NAME"] = $this->getRequestParam("BPARTNER_NAME");
    $RECORD["CREATEDBY"] = $this->getRequestParam("CREATEDBY");
    $RECORD["CREATEDON"] = $this->getRequestParam("CREATEDON");
    $RECORD["EMAIL"] = $this->getRequestParam("EMAIL");
    $RECORD["FAX"] = $this->getRequestParam("FAX");
    $RECORD["FIRSTNAME"] = $this->getRequestParam("FIRSTNAME");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["LASTNAME"] = $this->getRequestParam("LASTNAME");
    $RECORD["MOBILE"] = $this->getRequestParam("MOBILE");
    $RECORD["MODIFIEDBY"] = $this->getRequestParam("MODIFIEDBY");
    $RECORD["MODIFIEDON"] = $this->getRequestParam("MODIFIEDON");
    $RECORD["NAME"] = $this->getRequestParam("NAME");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["PHONE"] = $this->getRequestParam("PHONE");
    $sql = "insert into BP_CONTACT(
                 ID
                ,BPARTNER_ID
                ,NAME
                ,FIRSTNAME
                ,LASTNAME
                ,PHONE
                ,EMAIL
                ,FAX
                ,NOTES
                ,MOBILE
            ) values ( 
                 :ID
                ,:BPARTNER_ID
                ,:NAME
                ,:FIRSTNAME
                ,:LASTNAME
                ,:PHONE
                ,:EMAIL
                ,:FAX
                ,:NOTES
                ,:MOBILE
    )";
    $stmt = $this->db->prepare($sql);
    $this->logger->debug("insert of RECORD: ".$this->logger->map2string($RECORD) );
    $this->db->Execute($stmt, $RECORD);
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("end: ".$this->dcName.".doInsert");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doInsert */


public function doUpdate() {
  $this->logger->debug("Start: ".$this->dcName.".doUpdate");
  try {
    $RECORD = array();
    $RECORD["BPARTNER_ID"] = $this->getRequestParam("BPARTNER_ID");
    $RECORD["BPARTNER_NAME"] = $this->getRequestParam("BPARTNER_NAME");
    $RECORD["EMAIL"] = $this->getRequestParam("EMAIL");
    $RECORD["FAX"] = $this->getRequestParam("FAX");
    $RECORD["FIRSTNAME"] = $this->getRequestParam("FIRSTNAME");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["LASTNAME"] = $this->getRequestParam("LASTNAME");
    $RECORD["MOBILE"] = $this->getRequestParam("MOBILE");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["PHONE"] = $this->getRequestParam("PHONE");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0069.doUpdate().");}
    $sql = "update BP_CONTACT set 
                 ID=:ID
                ,BPARTNER_ID=:BPARTNER_ID
                ,FIRSTNAME=:FIRSTNAME
                ,LASTNAME=:LASTNAME
                ,PHONE=:PHONE
                ,EMAIL=:EMAIL
                ,FAX=:FAX
                ,NOTES=:NOTES
                ,MOBILE=:MOBILE
    where 
           ID= :ID
    ";
    $stmt = $this->db->prepare($sql);
    $this->logger->debug("update of RECORD: ".$this->logger->map2string($RECORD) );
    $this->db->Execute($stmt, $RECORD);
    $pkArray = array("ID" => $RECORD["ID"]);
    $this->findByPk($pkArray, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".doUpdate");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doUpdate */


public function doDelete() {
  $this->logger->debug("Start: ".$this->dcName.".doDelete");
  try {
    $RECORD["ID"] = $this->getRequestParam("ID");
    if (empty($RECORD["ID"])) { throw new Exception("Missing value for primary key field ID in DC0069.doDelete().");}
    $sql = "delete from BP_CONTACT where 
           ID= :ID
    ";
    $stmt = $this->db->prepare($sql);
    $this->db->Execute($stmt, $RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".doDelete");
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function doDelete */


public function initNewRecord() {
  try {
    $RECORD["BPARTNER_ID"] = $this->getRequestParam("BPARTNER_ID");
    $RECORD["BPARTNER_NAME"] = $this->getRequestParam("BPARTNER_NAME");
    $RECORD["CREATEDBY"] = $this->getRequestParam("CREATEDBY");
    $RECORD["CREATEDON"] = $this->getRequestParam("CREATEDON");
    $RECORD["EMAIL"] = $this->getRequestParam("EMAIL");
    $RECORD["FAX"] = $this->getRequestParam("FAX");
    $RECORD["FIRSTNAME"] = $this->getRequestParam("FIRSTNAME");
    $RECORD["ID"] = $this->getRequestParam("ID");
    $RECORD["LASTNAME"] = $this->getRequestParam("LASTNAME");
    $RECORD["MOBILE"] = $this->getRequestParam("MOBILE");
    $RECORD["MODIFIEDBY"] = $this->getRequestParam("MODIFIEDBY");
    $RECORD["MODIFIEDON"] = $this->getRequestParam("MODIFIEDON");
    $RECORD["NAME"] = $this->getRequestParam("NAME");
    $RECORD["NOTES"] = $this->getRequestParam("NOTES");
    $RECORD["PHONE"] = $this->getRequestParam("PHONE");
    $RECORD["_p_record_status"] = "insert";
    $this->setFieldInitialValues($RECORD, "DC0069");
    print "{success:true, data:".json_encode($RECORD)."}";
  }catch(Exception  $e) {
    System::sendActionErrorJson( $e->getMessage());
  }
} /* end function initNewRecord */

private function findByPk(&$pkCols, &$record) {
    $sql = "select 
                c.ID
                ,c.BPARTNER_ID
                ,c.NAME
                ,c.FIRSTNAME
                ,c.LASTNAME
                ,c.PHONE
                ,c.EMAIL
                ,c.FAX
                ,c.NOTES
                ,c.MOBILE
                ,c.CREATEDON
                ,c.CREATEDBY
                ,c.MODIFIEDON
                ,c.MODIFIEDBY
                ,(select bp.name from bpartner bp where bp.id = c.bpartner_id) BPARTNER_NAME
            from BP_CONTACT c
         where 
           c.ID= :ID
            ";
    $rs = $this->db->Execute($sql, $pkCols);
    $row = $rs->FetchRow();
    $this->copyArrayValues($record, $row);
} /* end function findByPk  */

private  $fieldDef = array(
  "ID" => array("DATA_TYPE" => "NUMBER")
  ,"BPARTNER_ID" => array("DATA_TYPE" => "NUMBER")
  ,"NAME" => array("DATA_TYPE" => "STRING")
  ,"FIRSTNAME" => array("DATA_TYPE" => "STRING")
  ,"LASTNAME" => array("DATA_TYPE" => "STRING")
  ,"PHONE" => array("DATA_TYPE" => "STRING")
  ,"EMAIL" => array("DATA_TYPE" => "STRING")
  ,"FAX" => array("DATA_TYPE" => "STRING")
  ,"NOTES" => array("DATA_TYPE" => "STRING")
  ,"MOBILE" => array("DATA_TYPE" => "STRING")
  ,"CREATEDON" => array("DATA_TYPE" => "DATE")
  ,"CREATEDBY" => array("DATA_TYPE" => "STRING")
  ,"MODIFIEDON" => array("DATA_TYPE" => "DATE")
  ,"MODIFIEDBY" => array("DATA_TYPE" => "STRING")
  ,"BPARTNER_NAME" => array("DATA_TYPE" => "STRING")
);


private function readRequest(&$RECORD) {
     if (isset($_REQUEST["BPARTNER_ID"] )) { $RECORD["BPARTNER_ID"] = $this->getRequestParam("BPARTNER_ID"); }
     if (isset($_REQUEST["BPARTNER_NAME"] )) { $RECORD["BPARTNER_NAME"] = $this->getRequestParam("BPARTNER_NAME"); }
     if (isset($_REQUEST["CREATEDBY"] )) { $RECORD["CREATEDBY"] = $this->getRequestParam("CREATEDBY"); }
     if (isset($_REQUEST["CREATEDON"] )) { $RECORD["CREATEDON"] = $this->getRequestParam("CREATEDON"); }
     if (isset($_REQUEST["EMAIL"] )) { $RECORD["EMAIL"] = $this->getRequestParam("EMAIL"); }
     if (isset($_REQUEST["FAX"] )) { $RECORD["FAX"] = $this->getRequestParam("FAX"); }
     if (isset($_REQUEST["FIRSTNAME"] )) { $RECORD["FIRSTNAME"] = $this->getRequestParam("FIRSTNAME"); }
     if (isset($_REQUEST["ID"] )) { $RECORD["ID"] = $this->getRequestParam("ID"); }
     if (isset($_REQUEST["LASTNAME"] )) { $RECORD["LASTNAME"] = $this->getRequestParam("LASTNAME"); }
     if (isset($_REQUEST["MOBILE"] )) { $RECORD["MOBILE"] = $this->getRequestParam("MOBILE"); }
     if (isset($_REQUEST["MODIFIEDBY"] )) { $RECORD["MODIFIEDBY"] = $this->getRequestParam("MODIFIEDBY"); }
     if (isset($_REQUEST["MODIFIEDON"] )) { $RECORD["MODIFIEDON"] = $this->getRequestParam("MODIFIEDON"); }
     if (isset($_REQUEST["NAME"] )) { $RECORD["NAME"] = $this->getRequestParam("NAME"); }
     if (isset($_REQUEST["NOTES"] )) { $RECORD["NOTES"] = $this->getRequestParam("NOTES"); }
     if (isset($_REQUEST["PHONE"] )) { $RECORD["PHONE"] = $this->getRequestParam("PHONE"); }
} /* end function readRequest  */


public function callProcedure($pName) {
  $this->logger->debug("Start: ".$this->dcName.".callProcedure(".$pName.")");
  try {
    $RECORD = array();
    $this->readRequest($RECORD);
    print "{success:true, data:".json_encode($RECORD)."}";
    $this->logger->debug("End: ".$this->dcName.".callProcedure");
  }catch(Exception  $e) {
      System::sendActionErrorJson( $e->getMessage());
  }
} /* end function callProcedure */

}
?>

File: ServerJava/src/net/nan21/ebs/dc/DC0069.java
/* N21 eBusiness Suite
 * Copyright: Nan21 Electronics srl
 * Generated content.
 * DC0069 DC Controller: Contacts
 */

package net.nan21.ebs.dc;


import java.sql.Connection;
import java.sql.ResultSet;
import java.util.Date;
import java.util.HashMap;
import java.util.Properties;
import javax.servlet.http.HttpServletResponse;
import net.nan21.ebs.lib.CollectionUtils;
import net.nan21.ebs.lib.AbstractDataControl;
import net.nan21.ebs.lib.FieldDef;
import net.nan21.ebs.lib.HttpRequest;
import net.nan21.ebs.lib.HttpSession;
import net.nan21.ebs.lib.IDataControl;
import net.nan21.ebs.lib.DbManager;

public class DC0069 extends AbstractDataControl implements IDataControl {

  public void init (HttpRequest request, HttpServletResponse response, HttpSession session, DbManager dbm) throws Exception {
    this._initFields();
    super.init(request, response, session, dbm);
  }

private void preQuery() {
    if (this.request.getParam("QRY_BPARTNER_ID") != null && !this.request.getParam("QRY_BPARTNER_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.BPARTNER_ID like :BPARTNER_ID");
      this.queryParams.put("BPARTNER_ID",(String)this.request.getParam("QRY_BPARTNER_ID"));
    }
    if (this.request.getParam("QRY_FIRSTNAME") != null && !this.request.getParam("QRY_FIRSTNAME").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.FIRSTNAME like :FIRSTNAME");
      this.queryParams.put("FIRSTNAME",(String)this.request.getParam("QRY_FIRSTNAME"));
    }
    if (this.request.getParam("QRY_ID") != null && !this.request.getParam("QRY_ID").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.ID like :ID");
      this.queryParams.put("ID",(String)this.request.getParam("QRY_ID"));
    }
    if (this.request.getParam("QRY_LASTNAME") != null && !this.request.getParam("QRY_LASTNAME").equals("")) {
      this.queryWhere.append(( this.queryWhere.length() > 0 )?" and ":"");
      this.queryWhere.append("c.LASTNAME like :LASTNAME");
      this.queryParams.put("LASTNAME",(String)this.request.getParam("QRY_LASTNAME"));
    }
}

public void doQuery() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " c.BPARTNER_ID"+
               " ,pbo_bpartner.get_name_by_id(c.bpartner_id) BPARTNER_NAME"+
               " ,c.CREATEDBY"+
               " ,to_char(c.CREATEDON,'"+this.DATE_FORMAT_DB+"') CREATEDON"+
               " ,c.EMAIL"+
               " ,c.FAX"+
               " ,c.FIRSTNAME"+
               " ,c.ID"+
               " ,c.LASTNAME"+
               " ,c.MOBILE"+
               " ,c.MODIFIEDBY"+
               " ,to_char(c.MODIFIEDON,'"+this.DATE_FORMAT_DB+"') MODIFIEDON"+
               " ,c.NAME"+
               " ,c.NOTES"+
               " ,c.PHONE"+
           " from BP_CONTACT c "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoQuery(sql);
} 


public void doExport() throws Exception {
    this.prepareQueryContext();
    this.preQuery();
    this.queryWhere.insert(0, (this.queryWhere.length()>0)?" where ":"");
    String sql = "select "+ 
               " c.ID"+
               " ,c.BPARTNER_ID"+
               " ,c.NAME"+
               ",pbo_bpartner.get_name_by_id(c.bpartner_id) BPARTNER_NAME"+
               " ,c.FIRSTNAME"+
               " ,c.LASTNAME"+
               " ,c.PHONE"+
               " ,c.EMAIL"+
               " ,c.FAX"+
               " ,c.NOTES"+
               " ,c.MOBILE"+
               " ,to_char(c.CREATEDON,'"+this.DATE_FORMAT_DB+"') CREATEDON"+
               " ,c.CREATEDBY"+
               " ,to_char(c.MODIFIEDON,'"+this.DATE_FORMAT_DB+"') MODIFIEDON"+
               " ,c.MODIFIEDBY"+
           " from BP_CONTACT c "+this.queryWhere.toString()+" "+this.queryOrderBy;
    this.writeResultDoExport(sql);
}

 	public void fetchRecord()  throws Exception  {
     this.populateRecordPkFromRequest();
     this.findByPk();
       this.writeResultFetchRecord();	 
   }

public void doInsert()  throws Exception {
  this.populateRecordFromRequest(); 
  this.populateRecordWithClientSpecific();
    String sql = "insert into BP_CONTACT("+
               "  BPARTNER_ID"+
               " ,EMAIL"+
               " ,FAX"+
               " ,FIRSTNAME"+
               " ,ID"+
               " ,LASTNAME"+
               " ,MOBILE"+
               " ,NAME"+
               " ,NOTES"+
               " ,PHONE"+
           " ) values ( "+
               "  :BPARTNER_ID"+
               " ,:EMAIL"+
               " ,:FAX"+
               " ,:FIRSTNAME"+
               " ,:ID"+
               " ,:LASTNAME"+
               " ,:MOBILE"+
               " ,:NAME"+
               " ,:NOTES"+
               " ,:PHONE"+
    ")";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoInsert();
}

public void doUpdate() throws Exception {
    this.populateRecordFromRequest();
    this.populateRecordWithClientSpecific();
    String sql = "update BP_CONTACT set "+
               "  ID=:ID"+
               " ,BPARTNER_ID=:BPARTNER_ID"+
               " ,FIRSTNAME=:FIRSTNAME"+
               " ,LASTNAME=:LASTNAME"+
               " ,PHONE=:PHONE"+
               " ,EMAIL=:EMAIL"+
               " ,FAX=:FAX"+
               " ,NOTES=:NOTES"+
               " ,MOBILE=:MOBILE"+
   " where "+
     "      ID= :ID"+
   "";
    dbm.executeStatement(sql, this.record);
    this.populateRecordPkFromRecord();
    this.findByPk();
    this.writeResultDoUpdate();	
}

public void doDelete() throws Exception {
    this.populateRecordPkFromRequest();
   String sql = "delete from BP_CONTACT where "+
     "      ID= :ID"+
    "";
    dbm.executeStatement(sql, this.recordPk);
   this.writeResultDoDelete();
}

public void initNewRecord() throws Exception {
    this.populateRecordFromRequest();
    this.record.put("_p_record_status", "insert");
    this.writeResultInitNewRecord();
}

private void findByPk()  throws Exception {
    String sql = "select "+ 
               " c.BPARTNER_ID"+
                ",pbo_bpartner.get_name_by_id(c.bpartner_id) BPARTNER_NAME"+
               " ,c.CREATEDBY"+
               " ,to_char(c.CREATEDON,'"+this.DATE_FORMAT_DB+"') CREATEDON"+
               " ,c.EMAIL"+
               " ,c.FAX"+
               " ,c.FIRSTNAME"+
               " ,c.ID"+
               " ,c.LASTNAME"+
               " ,c.MOBILE"+
               " ,c.MODIFIEDBY"+
               " ,to_char(c.MODIFIEDON,'"+this.DATE_FORMAT_DB+"') MODIFIEDON"+
               " ,c.NAME"+
               " ,c.NOTES"+
               " ,c.PHONE"+
           " from BP_CONTACT c"+
        " where "+
     "      c.ID= :ID"+ 
          "";
    this.record = dbm.executeQuery(sql, this.recordPk).get(0); 
} 


public void callProcedure(String pName)  throws Exception {
    this.populateRecordFromRequest();
}


	private void  _initFields() {
	  this.fields = new HashMap<String, FieldDef>();
	  this.fields.put("BPARTNER_ID", new FieldDef("NUMBER"));
	  this.fields.put("BPARTNER_NAME", new FieldDef("STRING"));
	  this.fields.put("CREATEDBY", new FieldDef("STRING"));
	  this.fields.put("CREATEDON", new FieldDef("DATE"));
	  this.fields.put("EMAIL", new FieldDef("STRING"));
	  this.fields.put("FAX", new FieldDef("STRING"));
	  this.fields.put("FIRSTNAME", new FieldDef("STRING"));
	  this.fields.put("ID", new FieldDef("NUMBER"));
	  this.fields.put("LASTNAME", new FieldDef("STRING"));
	  this.fields.put("MOBILE", new FieldDef("STRING"));
	  this.fields.put("MODIFIEDBY", new FieldDef("STRING"));
	  this.fields.put("MODIFIEDON", new FieldDef("DATE"));
	  this.fields.put("NAME", new FieldDef("STRING"));
	  this.fields.put("NOTES", new FieldDef("STRING"));
	  this.fields.put("PHONE", new FieldDef("STRING"));
	  String[] _pkFields = {"ID"};
	  this.pkFields = _pkFields;
	}

public void doCustomAction(String action) {}
}
